### PR TITLE
Refactor test collecting

### PR DESCRIPTION
**Pull request by cptartur**
Closes #

## Introduced changes

- This PR separates collecting tests with compiling test cases, so they can be tested separately. Also some of the code has been converted from free functions to struct methods.
- Replaced all occurrences of `&Utf8PathBuf` with `&Utf8Path`, same for `&PathBuf`.

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

File: crates/forge/src/lib.rs
Comment (piotmag769) [inline crates/forge/src/lib.rs:396]: commented out code

**Review by piotmag769 — APPROVED** [review]

**Comment by cptartur** [comment]
> I would like to review this one but after it is adapted to changes of #775

@MaksymilianDemitraszek it is ready for review now

**Review by piotmag769 — APPROVED** [review]

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:92]: Why this is done inside of the collecting function?

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:136]: Isn't it packing into one crate not file?

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:195]: Why this is tested here, not in the RunnerConfig file?

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:226]: collecting_test_compilation_targets

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:75]:
```suggestion
pub fn collect_test_compilation_targets(
```

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs:226]: Test cases:
How it behaves when there is no lib
How it behaves where there are no tests only lib

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:266]: Test case: what if names are duplicates? It can't happen because files in dir always have unique names, right?

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:531]: I don't understand what this test verifies

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:38]:
```suggestion
pub enum TestCrateLocation {
```

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:283]: What if has no tests?

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:168]: should be extracted to a separate function

**Review by MaksymilianDemitraszek — CHANGES_REQUESTED** [review]
Generally looks good, it is a positive change , It is ok to include it to the codebase after we remove cloning on large data, make it more idiomatic using iterators and improve names.

File: crates/forge/src/collecting.rs
Comment (piotmag769) [reply crates/forge/src/collecting.rs:30]: +1, changed it in #801

File: crates/forge/src/collecting.rs
Comment (piotmag769) [reply crates/forge/src/collecting.rs:134]: Would leave it, I extend this function in #886

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs:15]: Both names are a bit confusing, but it's not really a direct representation of a crate.

File: crates/forge/src/collecting.rs
Comment (piotmag769) [reply crates/forge/src/collecting.rs:226]: > How it behaves where there are no tests only lib

I don't think we need to test that, it is kinda testing the behaviour of the compiler

> How it behaves when there is no lib

It is already tested in e2e tests: `collection_with_lib` and `collection_without_lib` packages

File: crates/forge/src/collecting.rs
Comment (piotmag769) [reply crates/forge/src/collecting.rs:15]: What about `CompiledTestCrate`?

File: crates/forge/src/collecting.rs
Comment (piotmag769) [reply crates/forge/src/collecting.rs:92]: It is kinda a part of test collecting (if we want to return test crates from this function)

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs:18]: This way it's easier to test the filtering behavior, because we have a specific functions that can be called in unit tests. Otherwise we would have to test by calling a higher level function.

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs:226]: > Test cases: How it behaves when there is no lib 
> How it behaves where there are no tests only lib

so in both these cases the returned struct will contain a `TestCompilationTarget` representing a lib. It's hardcoded in the assumption that this always exists. 
The path to the lib is passed by the user as an argument, so I'm unsure if we should verify it's existence inside this method.

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs:531]: It was meant to verify that even for tests which somehow doesn't have a crate name attached (no part before `::`) the filtering logic works correctly. I'll combine it with other tests though.

File: crates/forge/src/collecting.rs
Comment (piotmag769) [reply crates/forge/src/collecting.rs:226]: Still imo e2e tests are enough there

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs:266]: So a test case for that doesn't make much sense

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs:283]: Tests stay an empty vector, I'll add an unit test for that.

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs]: I disagree, this is detail of compilation not collection, please rename the function or move it to compilation phase.

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs]: If you would implemented it idomatically with iterator you could possibly only test matching function.

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:16]: This is also a bad name. It suggest that it contains only compiled code. It also doesn't give an information of the scope, is this an arbitrary set of tests? Or tests from a module.  Please rename it to `TestCrate`

Comment (piotmag769) [reply crates/forge/src/collecting.rs:16]: `CompiledTestCrate`, we have `TestCrate` struct already

Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs:16]: It has been renamed.

Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs:16]: I am ok with: `TestCrate`, `CrateWithTests` or `CompiledTestCrate`, your call @cptartur

**Review by piotmag769 — APPROVED** [review]

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs]: > >   How it behaves where there are no tests only lib
> 
> I don't think we need to test that, it is kinda testing the behaviour of the compiler

I am not convinced, compiler doesn't know about test dir, we implement that on the top of it. How collection is going to behave if we call it on a dir without such directory?

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [reply crates/forge/src/collecting.rs]: I am not sure it this test actually test something useful. It is kinda testing logic of `name.contains` which is part of the std lib

**Review by MaksymilianDemitraszek — APPROVED** [review]
I am ok with merging this one after you remove packing tests from collecting phase

**Comment by cptartur** [comment]
> I am ok with merging this one after you remove packing tests from collecting phase

I'll do that

File: crates/forge/src/collecting.rs
Comment (cptartur) [reply crates/forge/src/collecting.rs]: Yeah, it's been removed, I've just extanded mocked_tests in other unit tests

File: crates/forge/src/collecting.rs
Comment (MaksymilianDemitraszek) [inline crates/forge/src/collecting.rs:17]:
```suggestion
    pub tests_source_location: CrateLocation,
```

File: crates/forge/src/collecting.rs
Comment (piotmag769) [inline crates/forge/src/collecting.rs:11]: todo

**Review by piotmag769 — APPROVED** [review]
We merge thoooose

**Comment by cptartur** [comment]
> We merge thoooose

Waiting for @MaksymilianDemitraszek re-review

**Review by MaksymilianDemitraszek — APPROVED** [review]
I would consider spliting test compilation target candidate (one without lib) and test compilation target instead of keeping it as a single type, imo this is a reason behind your problems in describing this contract, how `lib.cairo` is ensured is an implementation detail so there could be alternative strategies.  

Overall it improves state of the codebase good job, LGTM feel free to merge.

**Pull request merged**
